server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -642,14 +685,17 @@ private final class AsyncReplicaAction extends AbstractRunnable implements Actio
             this.replica = getIndexShard(shardId);
         }
 
+        protected abstract void shardOperation(ReplicaRequest request, ActionListener<ReplicaResult> listener);
+
+        public IndexShard getReplica() {
+            return this.replica;
+        }
+
         @Override
         public void onResponse(Releasable releasable) {
-            assert replica.getActiveOperationsCount() != 0 : "must perform shard operation under a permit";
             try {
-                shardOperationOnReplica(
-                    replicaRequest.getRequest(),
-                    replica,
-                    ActionListener.wrap((replicaResult) -> replicaResult.runPostReplicaActions(ActionListener.wrap(r -> {
+                ActionListener<ReplicaResult> listener = ActionListener.wrap(
+                    (replicaResult) -> replicaResult.runPostReplicaActions(ActionListener.wrap(r -> {
                         final ReplicaResponse response = new ReplicaResponse(
                             replica.getLocalCheckpoint(),
                             replica.getLastSyncedGlobalCheckpoint()
```

Comment:
I think we should not carry these checkpoints back from the unpromotable replicas. These checkpoints belong entirely to the world of promotable ones.